server/modules/playlist_driver.js

```diff
@@ -0,0 +1,111 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+'use strict';
+
+const monitor = require('server/monitoring/monitor');
+const random = require('random-js')();
+const time = require('server/util/time');
+const wallGeometry = require('server/util/wall_geometry');
+
+const driveStateMachine = (playlist, layoutSM) => {
+  const nextLayout = layoutIndex => {
+    // Show this layout next:
+    let layout = playlist[layoutIndex];
+    let partition = wallGeometry.partitionGeo(layout.maxPartitions);
+
+    // The time that we'll switch to a new layout.
+    let newLayoutTime = time.inFuture(layout.duration * 1000);
+
+    if (monitor.isEnabled()) {
+      monitor.update({playlist: {
+        time: time.now(),
+        event: `change layout`,
+        deadline: newLayoutTime
+      }});
+    }
+
+    layoutSM.setPartition(partition).then(() => {
+      // Shuffle the module list for each partition:
+      let modulesPerPartition = partition.map(() => {
+        let modules = Array.from(layout.modules);
+        random.shuffle(modules);
+        return modules;
+      });
+
+      const nextModule = moduleIndex => {
+        let timer;
+        layoutSM.setErrorListener(error => {
+          // Stop normal advancement.
+          clearTimeout(timer);
+          nextModule((moduleIndex + 1) % modulesPerPartition[0].length);
+        });
+
+        // The time that we'll switch to the next module.
+        let newModuleTime = time.inFuture(layout.moduleDuration * 1000);
+
+        // Tell each partition to play the next module in the list.
+        modulesPerPartition.forEach((modules, index) => {
+          layoutSM.playModule(index, modules[moduleIndex]);
+        });
+
+        if (monitor.isEnabled()) {
+          monitor.update({playlist: {
+            time: time.now(),
+            event: `change module`,
+            deadline: Math.min(newModuleTime, newLayoutTime)
+          }});
+        }
+
+        // Now, in so many seconds, we'll need to switch to another module or another layout. How much time do we
+        // have?
+        timer = (newLayoutTime < newModuleTime) ?
+          setTimeout(() => nextLayout((layoutIndex + 1) % playlist.length), time.until(newLayoutTime)) :
+          setTimeout(() => nextModule((moduleIndex + 1) % modulesPerPartition[0].length), time.until(newModuleTime));
+      };
+
+      Promise.all(layout.modules.map(m => m.whenLoadedPromise)).then(() => nextModule(0));
+    });
+  };
+  nextLayout(0);
+
+// Set up some byzantine failures. Here are the cases I want to test:
```

Comment:
can this block of comments and commented out code be moved out until you are ready to do something with it?

Reply:
Yes.